deepspeed/runtime/sequence_parallel/ulysses_sp.py

```diff
@@ -0,0 +1,1225 @@
+                torch.autograd.backward(output, incoming_grad_shard)
+
+        # positional args
+        grad_outputs = [None] * 9
+        # inject the grad for the position of forward input that is grad-requiring
+        arg_outputs = [None] * ctx.total_args
+        arg_outputs[grad_requiring_tensor_key_index] = grad_requiring_tensor_grad
+
+        return tuple(grad_outputs + arg_outputs)
+
+
+class TiledMLP(torch.autograd.Function):
+    """
+    Perform a tiled MLP computation to massively reduce memory usage needed to compute MLP when using very long sequence lengths
+
+    For a general tiled compute implementation that can handle any `forward` see `SequenceTiledCompute`
+
+    Args:
+    - fn: the function to call on sharded inputs
+    - `self`: the MLP nn.Module object
+    - `x`: the input to MLP.forward (`hidden_states`)
+    - `shards`: how many shards to use
+    - compute_params: a list of weights engaged in the compute Default: `None` (only needed when using DeepSpeed ZeRO)
+
+    Returns:
+    - the computed `hidden_states`
+
+    Here is an example that monkey patches HF Transformers' LLamaMLP:
+
+    def tiled_mlp_forward(self, x):
+        bs, seqlen, hidden = x.shape
+        num_shards = math.ceil(seqlen / hidden)
+        # to avoid deadlocks get all ranks to agree on the same num_shards by using the max value
+        tensor = torch.tensor(num_shards, device=x.device)
+        dist.all_reduce(tensor, op=dist.ReduceOp.MAX)
+        num_shards = tensor.item()
+        compute_params = [self.down_proj.weight, self.gate_proj.weight, self.up_proj.weight]
+
+        def mlp_forward(self, x):
+            return self.down_proj(self.act_fn(self.gate_proj(x)) * self.up_proj(x))
+
+        return TiledMLP.apply(
+            mlp_forward,
+            self,
+            x,
+            num_shards,
+            compute_params,
+        )
+
+        # this needs to be done before the model is instantiated
+        from transformers.models.llama import modeling_llama
+        modeling_llama.LlamaMLP.forward = tiled_mlp_forward
+    """
+
+    @staticmethod
+    def forward(
+        ctx,
+        fn,
+        self,
+        x,
+        shards,
+        compute_params,
+    ) -> torch.Tensor:
+        ctx.fn = fn
+        ctx.self = self
+        ctx.shards = shards
+        ctx.compute_params = compute_params
+        ctx.save_for_backward(x)
+
+        x_shards = list(torch.chunk(x, chunks=shards, dim=1))
+        with torch.no_grad():
+            output_shards = [fn(self, x_shard) for x_shard in x_shards]
+        output_unsharded = torch.cat(output_shards, dim=1)
+
+        return output_unsharded
+
+    @staticmethod
+    def backward(ctx, *grads) -> torch.Tensor:
+        fn = ctx.fn
+        (x, ) = ctx.saved_tensors
+        self = ctx.self
+        shards = ctx.shards
+        compute_params = ctx.compute_params
+        x_requires_grad = x.requires_grad
+        x = x.detach()
+        x.requires_grad_(x_requires_grad)
```

Comment:
The following sequence looks like a no-op. What does it do?

```suggestion
        x_requires_grad = x.requires_grad
        x = x.detach()
        x.requires_grad_(x_requires_grad)
```

Reply:
Please see:

https://github.com/pytorch/pytorch/blob/134179474539648ba7dee1317959529fbd0e7f89/torch/utils/checkpoint.py#L73-L74

yes/no/better way?

Reply:
here is why:

```
In [1]: import torch

In [2]: x = torch.zeros((1))

In [3]: x.requires_grad
Out[3]: False

In [4]: x = torch.zeros((1), requires_grad=True)

In [5]: x.requires_grad
Out[5]: True

In [6]: y = x.detach()

In [7]: y.requires_grad
Out[7]: False

In [8]: x = x.detach()

In [9]: x.requires_grad
Out[9]: False
```

as you can see `detach` drops `requires_grad` flag.

Reply:
so my code gets the original value and then restores it, probably could write it as:

```
x_requires_grad = x.requires_grad
x = x.detach().requires_grad_(x_requires_grad)
```
so it's more intuitive?